core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1188,119 +1133,201 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
+          val newBrokers = curBrokers.filter(broker => newBrokerIds(broker.id))
+          controllerContext.liveBrokers = curBrokers
+          val newBrokerIdsSorted = newBrokerIds.toSeq.sorted
+          val deadBrokerIdsSorted = deadBrokerIds.toSeq.sorted
+          val liveBrokerIdsSorted = curBrokerIds.toSeq.sorted
+          info("Newly added brokers: %s, deleted brokers: %s, all live brokers: %s"
+            .format(newBrokerIdsSorted.mkString(","), deadBrokerIdsSorted.mkString(","), liveBrokerIdsSorted.mkString(",")))
+          newBrokers.foreach(controllerContext.controllerChannelManager.addBroker)
+          deadBrokerIds.foreach(controllerContext.controllerChannelManager.removeBroker)
+          if(newBrokerIds.nonEmpty)
+            onBrokerStartup(newBrokerIdsSorted)
+          if(deadBrokerIds.nonEmpty)
+            onBrokerFailure(deadBrokerIdsSorted)
+        } catch {
+          case e: Throwable => error("Error while handling broker changes", e)
+        }
+      }
     }
-    partitionsToBeReassigned.foreach { partitionToBeReassigned =>
-      inLock(controllerContext.controllerLock) {
-        if(controller.deleteTopicManager.isTopicQueuedUpForDeletion(partitionToBeReassigned._1.topic)) {
-          error("Skipping reassignment of partition %s for topic %s since it is currently being deleted"
-            .format(partitionToBeReassigned._1, partitionToBeReassigned._1.topic))
-          controller.removePartitionFromReassignedPartitions(partitionToBeReassigned._1)
-        } else {
-          val context = new ReassignedPartitionsContext(partitionToBeReassigned._2)
-          controller.initiateReassignReplicasForTopicPartition(partitionToBeReassigned._1, context)
+  }
+
+  case class TopicChange(topics: Set[String]) extends ControllerEvent {
+    override def process(): Unit = {
+      if (!isActive) return
+      try {
+        val newTopics = topics -- controllerContext.allTopics
+        val deletedTopics = controllerContext.allTopics -- topics
+        controllerContext.allTopics = topics
+
+        val addedPartitionReplicaAssignment = zkUtils.getReplicaAssignmentForTopics(newTopics.toSeq)
+        controllerContext.partitionReplicaAssignment = controllerContext.partitionReplicaAssignment.filter(p =>
+          !deletedTopics.contains(p._1.topic))
+        controllerContext.partitionReplicaAssignment.++=(addedPartitionReplicaAssignment)
+        info("New topics: [%s], deleted topics: [%s], new partition replica assignment [%s]".format(newTopics,
+          deletedTopics, addedPartitionReplicaAssignment))
+        if (newTopics.nonEmpty)
+          onNewTopicCreation(newTopics, addedPartitionReplicaAssignment.keySet)
+      } catch {
+        case e: Throwable => error("Error while handling new topic", e)
+      }
+    }
+  }
+
+  case class PartitionModifications(topic: String) extends ControllerEvent {
+    override def process(): Unit = {
+      if (!isActive) return
```

Comment:
Is there a reason not to do the `!isActive` check before we call this method? We did something like that in the previous listener classes.